test/unittest/tests/extent_list.cxx

```diff
@@ -1,13 +1,393 @@
+#include <memory>
+
 #include <catch2/catch.hpp>
 
-uint32_t factorial(uint32_t num) {
-    return num > 1 ? factorial(num - 1) * num : 1;
+#include "wt_internal.h"
+
+struct ExtentWrapper {
+    ExtentWrapper(WT_EXT* raw)
+        : _raw(raw) {
+    }
+
+    ~ExtentWrapper() {
+        free(_raw);
+    }
+
+    WT_EXT *_raw;
+};
+
+struct ExtentListWrapper {
+    ExtentListWrapper() {
+    }
+
+    ~ExtentListWrapper() {
+    }
+
+    // unfortunately, some functions need raw 2D pointers, but that's not
+    // compatible with automatic memory management. _list is only for allocation
+    // bookkeeping - _raw_list can be rearranged however.
+    std::vector<std::unique_ptr<ExtentWrapper>> _list;
+    std::vector<WT_EXT*> _raw_list;
+};
+
+struct SizeWrapper {
+    SizeWrapper(WT_SIZE* raw)
+        : _raw(raw) {
+    }
+
+    ~SizeWrapper() {
+        free(_raw);
+    }
+
+    WT_SIZE *_raw;
+};
+
+struct SizeListWrapper {
+    SizeListWrapper() {
+    }
+
+    ~SizeListWrapper() {
+    }
+
+    // unfortunately, some functions need raw 2D pointers, but that's not
+    // compatible with automatic memory management. _list is only for allocation
+    // bookkeeping - _raw_list can be rearranged however.
+    std::vector<std::unique_ptr<SizeWrapper>> _list;
+    std::vector<WT_SIZE*> _raw_list;
+};
+
+std::unique_ptr<ExtentWrapper> create_new_ext() {
+    // manually alloc enough extra space for the zero-length array to encode two
+    // skip lists.
+    auto sz = sizeof(WT_EXT) + 2 * WT_SKIP_MAXDEPTH * sizeof(WT_EXT*);
+
+    auto raw = (WT_EXT*)malloc(sz);
+    memset(raw, 0, sz);
+
+    return std::make_unique<ExtentWrapper>(raw);
 }
 
-TEST_CASE("Factorials are computed", "[factorial]") {
-    REQUIRE(factorial(0) == 1);
-    REQUIRE(factorial(1) == 1);
-    REQUIRE(factorial(2) == 2);
-    REQUIRE(factorial(3) == 6);
-    REQUIRE(factorial(10) == 3628800);
+std::unique_ptr<SizeWrapper> create_new_sz() {
+    auto raw = (WT_SIZE*)malloc(sizeof(WT_SIZE));
+    memset(raw, 0, sizeof(WT_SIZE));
+
+    return std::make_unique<SizeWrapper>(raw);
+}
+
+void print_list(WT_EXT **head) {
+    WT_EXT* extp;
+    int i;
+
+    if (head == nullptr)
+        return;
+
+    for (i = 0; i < WT_SKIP_MAXDEPTH; i++) {
+        printf("L%d: ", i);
+
+        extp = head[i];
+        while (extp != nullptr) {
+            printf("%p -> ", extp);
+            extp = extp->next[i];
+        }
+
+        printf("X\n");
+    }
+}
+
+/*
+ * Creates a sane-looking "default" extent list suitable for testing:
+ * L0: 1 -> 2 -> 3 -> X
+ * L1: 2 -> 3 -> X
+ * L2: 3 -> X
+ * L3: X
+ * ...
+ * L9: X
+ */
+void create_default_test_extent_list(ExtentListWrapper& wrapper) {
+    auto& head = wrapper._list;
```

Comment:
I had a go at this, and it helps but there still has to be quite a bit of direct access anyway. This happens because when we're constructing the list internals we need to access the raw pointers to elements, so I ended up with code like:

```
void create_default_test_extent_list(ExtentListWrapper& wrapper) {
    for (int i = 0; i < 3; i++)
        wrapper.push_back(create_new_ext());
    for (int i = 3; i < WT_SKIP_MAXDEPTH; i++)
        wrapper.push_back(std::make_unique<ExtentWrapper>(nullptr));

    auto& head = wrapper._list;
    auto first = head[0]->_raw;
    auto second = head[1]->_raw;
    auto third = head[2]->_raw;
    first->next[0] = second;
    first->next[1] = third;
    second->next[0] = third;
}
```

which in my opinion isn't a huge improvement. It's possible I misunderstood you though?

Reply:
It would only be a small improvement, so I'm happy if we leave it for now and consider it in more detail later.